### PR TITLE
add missing Piglin Dancing API

### PR DESCRIPTION
**Pull request by Duckulus**
Add API to get and set the dancing state of Piglins.

https://minecraft.fandom.com/wiki/Piglin?file=Piglin_celebrate_hunt_Revision_2.gif

### CODE REVIEW

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
+@@ -100,5 +100,15 @@ public class CraftPiglin extends CraftPiglinAbstract implements Piglin, com.dest
+     public boolean isChargingCrossbow() {
+         return this.getHandle().isChargingCrossbow();
+     }
++
++    @Override
++    public void setDancing(boolean dancing) {
++        this.getHandle().setDancing(true);
```
Comment (electronicboy) [inline patches/server/0962-Piglin-Dancing-API.patch]: param derp

Comment (Duckulus) [reply patches/server/0962-Piglin-Dancing-API.patch]: oops

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
+index 48d0a4e42e1b90d1323784d1284acabfe9497dd6..c2d11d00eed5cb3c3c81e5e22af55d5318736e88 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
+@@ -1,9 +1,7 @@
+ package org.bukkit.craftbukkit.entity;
+ 
+ import com.google.common.base.Preconditions;
+-import java.util.Collections;
```
Comment (MiniDigger) [inline patches/server/0962-Piglin-Dancing-API.patch]: we generally don't touch imports in patches, so please revert the changes here and use the fully qualified name for the MemoryModuleType class

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
++import java.util.stream.Collectors;
++
+ public class CraftPiglin extends CraftPiglinAbstract implements Piglin, com.destroystokyo.paper.entity.CraftRangedEntity<net.minecraft.world.entity.monster.piglin.Piglin> { // Paper
+ 
+     public CraftPiglin(CraftServer server, net.minecraft.world.entity.monster.piglin.Piglin entity) {
+@@ -90,6 +92,7 @@ public class CraftPiglin extends CraftPiglinAbstract implements Piglin, com.dest
+     public String toString() {
+         return "CraftPiglin";
+     }
++
```
Comment (MiniDigger) [inline patches/server/0962-Piglin-Dancing-API.patch]: whitespace

File: patches/api/0418-Piglin-Dancing-API.patch [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+index d4cb4b0ed1d9766a87867dcf1a3a839526ba9332..44cbc24e1702161a17b21a151e945ab8b2b0a73f 100644
+--- a/src/main/java/org/bukkit/entity/Piglin.java
++++ b/src/main/java/org/bukkit/entity/Piglin.java
+@@ -109,6 +109,28 @@ public interface Piglin extends PiglinAbstract, InventoryHolder, com.destroystok
+      * @return is charging
+      */
+     boolean isChargingCrossbow();
++
++    /**
++     * Causes the piglin to dance
```
Comment (MiniDigger) [inline patches/api/0418-Piglin-Dancing-API.patch]: I am bad with words, but this sounds wrong. setDancing(false) doesn't cause the piglin to dance, it stops them from dancing. idk how to word it better tho

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+         return this.getHandle().isChargingCrossbow();
+     }
++
++    @Override
++    public void setDancing(boolean dancing) {
++        if (dancing) {
++            this.getHandle().getBrain().setMemory(MemoryModuleType.DANCING, true);
++            this.getHandle().getBrain().setMemory(MemoryModuleType.CELEBRATE_LOCATION, this.getHandle().getOnPos());
++        } else {
++            this.getHandle().getBrain().eraseMemory(MemoryModuleType.DANCING);
```
Comment (MiniDigger) [inline patches/server/0962-Piglin-Dancing-API.patch]: maybe we should erase celebrate location here too?

**Review by MiniDigger — CHANGES_REQUESTED** [review]
a few minor things

**Review by MiniDigger — APPROVED** [review]
fine by me now

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
++            this.getHandle().getBrain().setMemory(net.minecraft.world.entity.ai.memory.MemoryModuleType.CELEBRATE_LOCATION, this.getHandle().getOnPos());
++        } else {
++            this.getHandle().getBrain().eraseMemory(net.minecraft.world.entity.ai.memory.MemoryModuleType.DANCING);
++            this.getHandle().getBrain().eraseMemory(net.minecraft.world.entity.ai.memory.MemoryModuleType.CELEBRATE_LOCATION);
++        }
++    }
++
++    @Override
++    public void setDancing(int duration) {
++        this.getHandle().getBrain().setMemory(net.minecraft.world.entity.ai.memory.MemoryModuleType.DANCING, true);
```
Comment (MiniDigger) [inline patches/server/0962-Piglin-Dancing-API.patch]: actually, one more thing, I know nothing about the brain api, but why do you set one with an expiry and one without?

**Review by Owen1212055 — DISMISSED** [review]
Does the memory type api not cover this? I know MM has a PR to expose more memory types.

**Comment by yannicklamprecht** [comment]
> Does the memory type api not cover this? I know MM has a PR to expose more memory types.
Nope that isn't covered currently.

Probably a good thing to have a folowup mr to migrate this api impl to memory api calls.

File: patches/server/0962-Piglin-Dancing-API.patch (hunk 6 above)
Comment (Duckulus) [reply patches/server/0962-Piglin-Dancing-API.patch]: There isn't a particular reason for that, I changed it now so that both memories expire

File: patches/server/0962-Piglin-Dancing-API.patch [hunk 7]
```diff
@@ -26,7 +26,7 @@ index 48d0a4e42e1b90d1323784d1284acabfe9497dd6..b3b0d58082d39575788d93ba7c787576
 +
 +    @Override
 +    public void setDancing(int duration) {
```
Comment (molor) [inline patches/server/0962-Piglin-Dancing-API.patch]: Maybe change the type to `long`? iirc the `setMemoryWithExpiry` method also accepts a `long` value, and bunch of methods in the API that accepts ticks uses `long`

**Review by kennytv — APPROVED** [review]

**Review by yannicklamprecht — APPROVED** [review]

**Comment by lynxplay** [comment]
This should go into the missing entity API patches, beyond that, LGTM.

**Review by lynxplay — APPROVED** [review]
Merged the two patches and added you as a co-author to the missing entity api patch :+1: LGTM
